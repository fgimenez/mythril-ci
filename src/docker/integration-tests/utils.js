import config from 'config';
import httpStatus from 'http-status';
import {MongoClient} from 'mongodb';
import request from 'supertest';
import uuidV4 from 'uuid/v4';

const serverRequest = request(`http://127.0.0.1:${process.env.PORT}`);

/**
 * For the given uuid, waits until the current status is no longer the passed currentStatus argument, and expects
 * the new status is the passed nextStatus argument.
 * @param {string} uuid
 * @param {string} currentStatus
 * @param {string} nextStatus
 * @param {string} token
 * @param {function} expect
 */
async function waitForStatusUpdate(uuid, currentStatus, nextStatus, token, expect) {
  let res;

  // eslint-disable-next-line no-constant-condition, no-restricted-syntax
  while (true) {
    const delayMS = 50;
    await Promise.delay(delayMS);

    res = await serverRequest
      .get(`/mythril/v1/analysis/${uuid}`)
      .set('Authorization', `Bearer ${token}`)
      .expect(httpStatus.OK);
    const result = res.body.result;
    if (result !== currentStatus) {
      expect(result).toBe(nextStatus);
      break;
    }
  }
}

/**
 * Generate random email address
 * @returns {string} random email address
 */
function generateEmailAddress() {
  return `${uuidV4()}@test.com`;
}

/**
 * Fetch user object from database
 * @param {string} email email address
 * @returns {string} random email address
 */
async function getUserFromDatabase(email) {
  const client = await MongoClient.connect(
    config.MONGODB_URL, {
      useNewUrlParser: true,
    },
  );
  let user;
  try {
    const userCollection = await client.db().collection('users');
    user = await userCollection.findOne({email_lowered: email.toLowerCase()});
  } finally {
    await client.logout();
  }
  return user;
}

/**
 * Get valid credential
 *
 * @returns {object} email and token
 */
async function getValidCredential() {
  const email = generateEmailAddress();
  await serverRequest
    .post('/mythril/v1/auth/user')
    .send({
      firstName: 'David',
      lastName: 'Martin',
      email,
      termsId: 'no_terms',
    });
  const user = await getUserFromDatabase(email);
  return {email, token: user.apiKey};
}

/**
 * Get valid user
 *
 * @returns {object} user
 */
async function getValidUser() {
  const {email} = await getValidCredential();
  return await getUserFromDatabase(email);
}

/**
 * Set user property
 * @param {string} email email address
 * @param {object} values
 * @returns {object} user user object
 */
async function setUserProperty(email, values) {
  const client = await MongoClient.connect(
    config.MONGODB_URL, {
      useNewUrlParser: true,
    },
  );
  let user;
  try {
    const userCollection = await client.db().collection('users');
    user = await userCollection.findOneAndUpdate({email_lowered: email.toLowerCase()}, {$set: values});
  } finally {
    await client.logout();
  }
  return user;
}

/**
 * Make user type unlimited
 * @param {string} email email address
 * @param {object} values
 */
async function makeUserUnlimited(email) {
  await setUserProperty(email, {type: 'unlimited'});
}

export {
  serverRequest,
  waitForStatusUpdate,
  generateEmailAddress,
  getValidCredential,
  getUserFromDatabase,
  getValidUser,
  makeUserUnlimited,
  setUserProperty,
};
