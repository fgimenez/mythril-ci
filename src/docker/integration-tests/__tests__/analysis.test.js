import _ from 'lodash';
import httpStatus from 'http-status';
import {serverRequest, makeUserUnlimited, waitForStatusUpdate, getValidCredential} from '../utils';
import submissionWithIssues from './submissionWithIssues';

describe('/mythril/v1/analysis', () => {
  describe('Submit', () => {
    it('post analysis without authorization', async () => {
      const res = await serverRequest
        .post('/mythril/v1/analysis')
        .send({
          type: 'bytecode',
          contract: 'abcc',
        })
        .expect(httpStatus.UNAUTHORIZED);

      expect(res.body.status).toBe(httpStatus.UNAUTHORIZED);
    });

    it('get analysis status without authorization', async () => {
      const res = await serverRequest
        .get('/mythril/v1/analysis/notexist')
        .expect(httpStatus.UNAUTHORIZED);

      expect(res.body.status).toBe(httpStatus.UNAUTHORIZED);
    });

    it('get analysis issues without authorization', async () => {
      const res = await serverRequest
        .get('/mythril/v1/analysis/notexist/issues')
        .expect(httpStatus.UNAUTHORIZED);

      expect(res.body.status).toBe(httpStatus.UNAUTHORIZED);
    });

    it('no issues', async () => {
      const {email, token} = await getValidCredential();
      await makeUserUnlimited(email);

      let res = await serverRequest
        .post('/mythril/v1/analysis')
        .set('Authorization', `Bearer ${token}`)
        .send({
          type: 'bytecode',
          contract: 'abcc',
        })
        .expect(httpStatus.OK);

      expect(res.body.result).toBe('Queued');
      expect(res.body).toHaveProperty('uuid');

      await waitForStatusUpdate(res.body.uuid, 'Queued', 'In progress', token, expect);
      await waitForStatusUpdate(res.body.uuid, 'In progress', 'Finished', token, expect);

      res = await serverRequest
        .get(`/mythril/v1/analysis/${res.body.uuid}/issues`)
        .set('Authorization', `Bearer ${token}`)
        .expect(httpStatus.OK);

      const foundIssues = res.body;
      expect(foundIssues).toBeInstanceOf(Array);
      expect(foundIssues.length).toBe(0);
    });

    it('error', async () => {
      const {email, token} = await getValidCredential();
      await makeUserUnlimited(email);

      const res = await serverRequest
        .post('/mythril/v1/analysis')
        .set('Authorization', `Bearer ${token}`)
        .send({
          type: 'bytecode',
          contract: '01',
        })
        .expect(httpStatus.OK);
      expect(res.body.result).toBe('Queued');
      expect(res.body).toHaveProperty('uuid');

      await waitForStatusUpdate(res.body.uuid, 'Queued', 'In progress', token, expect);
      await waitForStatusUpdate(res.body.uuid, 'In progress', 'Error', token, expect);
    });

    it('Submit multiple (no issues)', async () => {
      const {email, token} = await getValidCredential();
      await makeUserUnlimited(email);

      let res = await serverRequest
        .post('/mythril/v1/analysis')
        .set('Authorization', `Bearer ${token}`)
        .send({
          type: 'bytecode',
          contracts: ['abcc', '00', '11'],
        })
        .expect(httpStatus.OK);
      expect(res.body.result).toBe('Queued');
      expect(res.body).toHaveProperty('uuid');

      await waitForStatusUpdate(res.body.uuid, 'Queued', 'In progress', token, expect);
      await waitForStatusUpdate(res.body.uuid, 'In progress', 'Finished', token, expect);

      res = await serverRequest
        .get(`/mythril/v1/analysis/${res.body.uuid}/issues`)
        .set('Authorization', `Bearer ${token}`)
        .expect(httpStatus.OK);

      const foundIssues = res.body;
      expect(foundIssues).toBeInstanceOf(Array);
      expect(foundIssues.length).toBe(0);
    });

    it('issues', async () => {
      const {email, token} = await getValidCredential();
      await makeUserUnlimited(email);

      let res = await serverRequest
        .post('/mythril/v1/analysis')
        .set('Authorization', `Bearer ${token}`)
        .send(submissionWithIssues)
        .expect(httpStatus.OK);

      expect(res.body.result).toBe('Queued');
      expect(res.body).toHaveProperty('uuid');

      await waitForStatusUpdate(res.body.uuid, 'Queued', 'In progress', token, expect);
      await waitForStatusUpdate(res.body.uuid, 'In progress', 'Finished', token, expect);

      res = await serverRequest
        .get(`/mythril/v1/analysis/${res.body.uuid}/issues`)
        .set('Authorization', `Bearer ${token}`)
        .expect(httpStatus.OK);

      const issues = res.body.map((issue) => _.omit(issue, 'debug'));

      expect(issues).toMatchSnapshot();
    });
  });
});
