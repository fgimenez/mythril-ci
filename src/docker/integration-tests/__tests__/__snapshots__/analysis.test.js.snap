// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/mythril/v1/analysis Submit issues 1`] = `
Array [
  Object {
    "address": 499,
    "contract": "MAIN",
    "debug": "callvalue: 0x57f25cd000000000000000000000000000000000000000000000000000000000
storage_keccac_1461501637330902918203684832716283019655932542975_&
1461501637330902918203684832716283019655932542975_&
1461501637330902918203684832716283019655932542975_&
calldata_MAIN[4]: 0xa8795cd3fffffffffffffffffffffffffffffffffffffffffffffffffffffffe
calldata_MAIN[0]: 0x362a9500000000000000000000000000000000000000000000000000000000
calldatasize_MAIN: 0x4
",
    "description": "A possible integer overflow exists in the function \`fallback()\`.
The addition or multiplication may result in a value higher than the maximum representable integer.",
    "function": "fallback()",
    "title": "Integer Overflow",
    "type": "Warning",
  },
  Object {
    "address": 648,
    "contract": "MAIN",
    "debug": "",
    "description": "This contract executes a message call to the address of the transaction sender. Generally, it is not recommended to call user-supplied addresses using Solidity's call() construct. Note that attackers might leverage reentrancy attacks to exploit race conditions or manipulate this contract's state.",
    "function": "_function_0x2e1a7d4d",
    "title": "Message call to external contract",
    "type": "Warning",
  },
  Object {
    "address": 648,
    "contract": "MAIN",
    "debug": "",
    "description": "Multiple sends exist in one transaction, try to isolate each external call into its own transaction. As external calls can fail accidentally or deliberately.
Consecutive calls: 
Call at address: 823
",
    "function": "_function_0x2e1a7d4d",
    "title": "Multiple Calls",
    "type": "Information",
  },
  Object {
    "address": 728,
    "contract": "MAIN",
    "debug": "",
    "description": "The contract account state is changed after an external call. Consider that the called contract could re-enter the function before this state change takes place. This can lead to business logic vulnerabilities.",
    "function": "_function_0x2e1a7d4d",
    "title": "State change after external call",
    "type": "Warning",
  },
  Object {
    "address": 823,
    "contract": "MAIN",
    "debug": "",
    "description": "This contract executes a message call to to another contract. Make sure that the called contract is trusted and does not execute user-supplied code.",
    "function": "_function_0x2e1a7d4d",
    "title": "Message call to external contract",
    "type": "Informational",
  },
  Object {
    "address": 823,
    "contract": "MAIN",
    "debug": "",
    "description": "The return value of an external call is not checked. Note that execution continue even if the called contract throws.",
    "function": "_function_0x2e1a7d4d",
    "title": "Unchecked CALL return value",
    "type": "Informational",
  },
  Object {
    "address": 829,
    "contract": "MAIN",
    "debug": "",
    "description": "The return value of an external call is not checked. Note that execution continue even if the called contract throws.",
    "function": "_function_0x2e1a7d4d",
    "title": "Unchecked CALL return value",
    "type": "Informational",
  },
]
`;
